### PR TITLE
[APITESTS] Fix some tests

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Fix some tests to pass on WHS

## Proposed changes

- [TESTS] Instead of entirely skipping tests with ros_skip_flaky, only disable counting errors
- [APITESTS] Add PROJECT(ApiTests) to get a separate VS solution file
- [NTDLL_APITEST] Remove duplicated test for RtlInitilizeBitMap
- [GDI32_APITEST] Remove duplicated test in Test_CreateBitmap
- [PSDK] Add extern "C" for c++ code in mmddk.h
- [PSDK] Add versionhelpers.h
- [APITESTS] Improve and fix tests

## Test results
* https://reactos.org/testman/compare.php?ids=70709,70722

### CODE REVIEW

File: modules/rostests/apitests/gdi32/CombineTransform.c [hunk 1]
```diff
@@ -45,9 +45,48 @@ float geINF;
...
+    /* Different behavior between native and Wow64 */
+    if (IsWow64Process(GetCurrentProcess(), &IsWow64) && IsWow64)
+    {
+        eTooBigValue = 4294967296.0f;
+    }
+    else
+    {
+        eTooBigValue = 4294967296.0f;
+    }
+printf("eTooBigValue = %f\n", eTooBigValue);
```
Comment (learn-more) [inline modules/rostests/apitests/gdi32/CombineTransform.c]: nit: `trace`?

File: modules/rostests/apitests/CMakeLists.txt [hunk 2]
```diff
@@ -1,3 +1,6 @@
+
+PROJECT(ApiTests)
```
Comment (learn-more) [inline modules/rostests/apitests/CMakeLists.txt]: This is a good idea, but you might need to build them all in the full solution first to get all dependencies built

File: modules/rostests/apitests/gdi32/CreateBitmap.c [hunk 3]
```diff
@@ -136,32 +135,6 @@ void Test_CreateBitmap()
     hbmp = CreateBitmap(1, 2, 1, 33, NULL);
     ok(hbmp == 0, "should fail\n");
 
-    for (TestBitsPixel = 0; TestBitsPixel <= 32; TestBitsPixel++)
```
Comment (learn-more) [inline modules/rostests/apitests/gdi32/CreateBitmap.c]: Flaky or?

File: modules/rostests/apitests/ntdll/CMakeLists.txt [hunk 4]
```diff
@@ -56,7 +56,6 @@ list(APPEND SOURCE
     RtlGetLongestNtPathLength.c
     RtlHandle.c
     RtlImageRvaToVa.c
-    RtlInitializeBitMap.c
```
Comment (learn-more) [inline modules/rostests/apitests/ntdll/CMakeLists.txt]: What is wrong with this test?

Comment (HBelusca) [reply modules/rostests/apitests/ntdll/CMakeLists.txt]: It's the redundant test that is already covered with RtlBitmap.c in that same apitest module.

File: modules/rostests/apitests/gdi32/CreateBitmap.c (hunk 3 above)
Comment (tkreuzer) [reply modules/rostests/apitests/gdi32/CreateBitmap.c]: It's simply the same code as below.

File: modules/rostests/apitests/atl/CImage.cpp [hunk 5]
```diff
@@ -208,7 +208,7 @@ START_TEST(CImage)
     height = image2.GetHeight();
     ok(height == 48, "Expected height to be 48, was: %d\n", height);
     bpp = image2.GetBPP();
-    ok(bpp == 8, "Expected bpp to be 8, was: %d\n", bpp);
+    ok(bpp == 32, "Expected bpp to be 8, was: %d\n", bpp);
```
Comment (learn-more) [inline modules/rostests/apitests/atl/CImage.cpp]: can you please update the text indicating 8 as well? (same for the next 2)

File: modules/rostests/apitests/shell32/IShellFolderViewCB.cpp [hunk 6]
```diff
@@ -246,11 +246,15 @@ class CFolderViewCB :
     public IShellFolderViewCB
 {
 public:
+    CFolderViewCB(void) :
+        m_RefCount(1)
```
Comment (learn-more) [inline modules/rostests/apitests/shell32/IShellFolderViewCB.cpp]: Why did you add a refcount here?

File: sdk/include/reactos/wine/test.h [hunk 7]
```diff
@@ -548,25 +554,25 @@ void winetest_wait_child_process( HANDLE process )
     }
 }
 
-void winetest_disable_success_count_start()
+void winetest_nocount_start()
 {
-    get_tls_data()->nocount_level <<= 1;
+    get_tls_data()->nocount_level <<= 2;
```
Comment (learn-more) [inline sdk/include/reactos/wine/test.h]: Why is this a bit shift instead of just adding / subtracting?

File: modules/rostests/apitests/shell32/IShellFolderViewCB.cpp (hunk 6 above)
Comment (tkreuzer) [reply modules/rostests/apitests/shell32/IShellFolderViewCB.cpp]: To make it behave like a refcounter. Maybe I'm getting something wrong here, but IIUC this is not a callback, but an actual reference counter for the callback object. I don't really understand the referencing behavior (like why it's not dereferenced, when a new callback object is set), but this seems to make much more sense.

Comment (learn-more) [reply modules/rostests/apitests/shell32/IShellFolderViewCB.cpp]: The object is a callback, which has a reference counter,
but since I didnt want to have it dynamically destruct I left it at a fixed reference count.

File: sdk/include/reactos/wine/test.h (hunk 7 above)
Comment (tkreuzer) [reply sdk/include/reactos/wine/test.h]: It was a workaround for making this work with a for loop. The loop exit condition is checked before the code in question is run. With a simple counter you cannot check, whether you already ran the code once. With the shift you have different states: 1. "initialized" (which shifts the bits and sets the lowest 2 bits to 0, 2. "set", which sets the disable bit as intended and tells the loop condition to abort next time, and 3. "reset", which is back to where it was before the loop. After adding a second bit, I introduced a bug though: if one bit is already set and then the other bit is added, the previous bit is lost. I'll gix that.

Comment (HBelusca) [reply sdk/include/reactos/wine/test.h]: add a comment? (nobody knew about that idea except you :P)

File: modules/rostests/apitests/atl/CImage.cpp (hunk 5 above)
Comment (tkreuzer) [reply modules/rostests/apitests/atl/CImage.cpp]: Done

File: modules/rostests/apitests/shell32/IShellFolderViewCB.cpp (hunk 6 above)
Comment (tkreuzer) [reply modules/rostests/apitests/shell32/IShellFolderViewCB.cpp]: But the static counters were not reference counters, but call counters.

File: sdk/include/reactos/wine/test.h (hunk 7 above)
Comment (tkreuzer) [reply sdk/include/reactos/wine/test.h]: ok :)

File: modules/rostests/apitests/shell32/ShellState.cpp [hunk 8]
```diff
@@ -11,6 +11,7 @@
 #include <stdio.h>
 #include <shellutils.h>
 #include <strsafe.h>
+#include <Shlwapi.h>
```
Comment (learn-more) [inline modules/rostests/apitests/shell32/ShellState.cpp]: Should be lowercase `s` for linux machines

File: modules/rostests/apitests/gdi32/CreateBitmap.c [hunk 9]
```diff
@@ -189,8 +162,8 @@ void Test_CreateBitmap()
         ok(hbmp != 0, "should get a 1x2 bitmap %ld\n", cBitsPixel);
         ok_int(GetObject(hbmp, sizeof(bitmap), &bitmap), sizeof(BITMAP));
 
-        /* calculate expected line width */
-        cjWidthBytes = ((((ULONG)bitmap.bmWidth) * ((ULONG)bitmap.bmBitsPixel) + 15) & ~15) >> 3;
+        /* Calculate expected line width */
+        cjWidthBytes = ((bitmap.bmWidth * bitmap.bmBitsPixel + 15) & ~15) >> 3;
```
Comment (ThFabba) [inline modules/rostests/apitests/gdi32/CreateBitmap.c]: I can't tell if this is cosmetic or to fix a test failure.

File: modules/rostests/apitests/gdi32/CreateDIBitmap.c [hunk 10]
```diff
@@ -334,7 +334,7 @@ Test_CreateDIBitmap1(void)
     ok(bitmap.bmType == 0, "\n");
     ok(bitmap.bmWidth == 2, "\n");
     ok(bitmap.bmHeight == 2, "\n");
-    ok(bitmap.bmWidthBytes == 8, "bmWidthBytes = %ld\n", bitmap.bmWidthBytes);
+    ok(bitmap.bmWidthBytes == 4, "bmWidthBytes = %ld\n", bitmap.bmWidthBytes);
```
Comment (ThFabba) [inline modules/rostests/apitests/gdi32/CreateDIBitmap.c]: Should this be `bitmap.bmWidth * 8 *  bitmap.bmBitsPixel` or some variation thereof?

File: modules/rostests/apitests/gdi32/CreateIconIndirect.c [hunk 11]
```diff
@@ -147,45 +145,45 @@ START_TEST(CreateIconIndirect)
     ok(hcursor != 0, "should not fail\n");
     ok(GetIconInfo(hcursor, &iconinfo2), "\n");
     ok(iconinfo2.fIcon == 0, "\n");
-    ok(iconinfo2.xHotspot == 0, "%ld\n", iconinfo2.xHotspot);
-    ok(iconinfo2.yHotspot == 8, "%ld\n", iconinfo2.yHotspot);
+    ok_int(iconinfo2.xHotspot, 10);
+    ok_int(iconinfo2.yHotspot, 10);
```
Comment (ThFabba) [inline modules/rostests/apitests/gdi32/CreateIconIndirect.c]: It seems silly to even test this. It depends on whatever your system's cursor is set to, doesn't it?

I'd argue that's true for almost everything between L146/144 to L171/169.

File: modules/rostests/apitests/user32/WndProc.c [hunk 12]
```diff
@@ -25,15 +25,12 @@ static LRESULT WINAPI redraw_window_procA(
...
-    DWORD ExtraData,
-    DWORD ExtraData1,
-    DWORD ExtraData2,
-    DWORD ExtraData3)
+    LPARAM lparam)
 {
     switch (msg)
     {
         case WM_PAINT:
+            break;
```
Comment (ThFabba) [inline modules/rostests/apitests/user32/WndProc.c]: Strange hack.

File: modules/rostests/apitests/shell32/ShellState.cpp (hunk 8 above)
Comment (tkreuzer) [reply modules/rostests/apitests/shell32/ShellState.cpp]: fixed

File: sdk/include/reactos/wine/test.h (hunk 7 above)
Comment (tkreuzer) [reply sdk/include/reactos/wine/test.h]: I changed it to use the same structure as todo_wine, this should make things better understandable. I also added a comment.

File: modules/rostests/apitests/browseui/SHExplorerParseCmdLine.c [hunk 13]
```diff
@@ -261,8 +262,8 @@ START_TEST(SHExplorerParseCmdLine)
         { __LINE__, L"::{645ff040-5081-101b-9f08-00aa002f954e}", TRUE, PIDL_IS_PATH, 0x00000200, NULL, L"::{645ff040-5081-101b-9f08-00aa002f954e}" },
         { __LINE__, L"/select,::{450d8fba-ad25-11d0-98a8-0800361b1103}", TRUE, PIDL_IS_PATH, 0x00000240, NULL, L"::{450d8fba-ad25-11d0-98a8-0800361b1103}" },
         { __LINE__, L"=", TRUE, PIDL_IS_PATH, 0x00000200, NULL, L"::{20d04fe0-3aea-1069-a2d8-08002b30309d}" },
-        { __LINE__, L".", TRUE, PIDL_IS_PATH, 0x00000200, NULL, L"C:\\Documents and Settings\\gigaherz\\Desktop" },
-        { __LINE__, L"..", TRUE, PIDL_IS_PATH, 0x00000200, NULL, L"C:\\Documents and Settings\\gigaherz" },
+//        { __LINE__, L".", TRUE, PIDL_IS_PATH, 0x00000200, NULL, L"C:\\Documents and Settings\\gigaherz\\Desktop" },
+//        { __LINE__, L"..", TRUE, PIDL_IS_PATH, 0x00000200, NULL, L"C:\\Documents and Settings\\gigaherz" },
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply modules/rostests/apitests/browseui/SHExplorerParseCmdLine.c]: One problem is that they contain the user name ("gigaherz"). But that wasn't really the failure. The failure is:
```SHExplorerParseCmdLine.c:181: Test failed: Unexpected pidl value 0024AA10; pidlPathName=C:\WINDOWS pidlPathTest=C:\Documents and Settings\gigaherz\Desktop```
```SHExplorerParseCmdLine.c:181: Test failed: Unexpected pidl value 0024E5A0; pidlPathName=C:\ pidlPathTest=C:\Documents and Settings\gigaherz  ```
No idea, what's going on there.

File: modules/rostests/apitests/comctl32/button.c [hunk 14]
```diff
@@ -661,6 +661,7 @@ MSG_ENTRY enable_nonthemed_sequence[]={
 MSG_ENTRY btndown_nonthemed_sequence[]={
     {2, WM_LBUTTONDOWN, SENT},
     {1, WM_KILLFOCUS, SENT},
+    {1, WM_IME_SETCONTEXT, SENT},
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply modules/rostests/apitests/comctl32/button.c]: All I know is, that this works. I leave that to somebody who understands the code better.

File: modules/rostests/apitests/gdi32/CreateBitmap.c (hunk 9 above)
Comment (tkreuzer) [reply modules/rostests/apitests/gdi32/CreateBitmap.c]: Cosmetic.

File: modules/rostests/apitests/gdi32/CreateDIBitmap.c (hunk 10 above)
Comment (tkreuzer) [reply modules/rostests/apitests/gdi32/CreateDIBitmap.c]: Should be something like ```ALIGN_UP_BY(bitmap.bmWidth * bitmap.bmBitsPixel, 16) / 8``` I'll fix that.

File: modules/rostests/apitests/gdi32/CreateIconIndirect.c (hunk 11 above)
Comment (tkreuzer) [reply modules/rostests/apitests/gdi32/CreateIconIndirect.c]: I'll improve it a bit. Most of that stuff shouldn't be here in the first place, but fixing that is not part of this PR.

File: modules/rostests/apitests/user32/WndProc.c (hunk 12 above)
Comment (tkreuzer) [reply modules/rostests/apitests/user32/WndProc.c]: The whole test was (is?) a huge hack. At least now it doesn't crash anymore.

File: modules/rostests/apitests/browseui/SHExplorerParseCmdLine.c [hunk 15]
```diff
@@ -97,34 +98,34 @@ _Out_opt_ PUINT PWriteEnd)
...
 
-    ok(Info.dwFlags == ExpectedFlags, "dwFlags = %08lx, expected %08lx\n", Info.dwFlags, ExpectedFlags);
+    ok(Info.dwFlags == ExpectedFlags, "Line %lu: dwFlags = %08lx, expected %08lx\n", TestLine, Info.dwFlags, ExpectedFlags);
 
     if (ExpectedCsidl == PIDL_IS_UNTOUCHED)
-        ok(Info.pidl == InvalidPointer, "pidl = %p\n", Info.pidl);
+        ok(Info.pidl == InvalidPointer, "Line %lu: pidl = %p\n", TestLine, Info.pidl);
     else if (ExpectedCsidl == PIDL_IS_NULL)
-        ok(Info.pidl == NULL, "pidl = %p\n", Info.pidl);
+        ok(Info.pidl == NULL, "Line %lu:  = %p\n", TestLine, Info.pidl);
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply modules/rostests/apitests/browseui/SHExplorerParseCmdLine.c]: fixed

File: modules/rostests/apitests/comctl32/button.c [hunk 16]
```diff
@@ -795,31 +796,31 @@ void Test_MessagesNonThemed()
     COMPARE_CACHE(btndown_nonthemed_sequence);
 
     state = SendMessageW(hWnd2, BM_GETSTATE,0,0);
-    ok(state == 0, "Expected state 0, got %lu\n", state);
+    ok(state == 108, "Expected state 0, got %lu\n", state);
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply modules/rostests/apitests/comctl32/button.c]: Fixed

File: modules/rostests/apitests/comctl32/button.c [hunk 17]
```diff
@@ -795,31 +796,31 @@ void Test_MessagesNonThemed()
...
     EMPTY_CACHE();
 
     SendMessageW(hWnd2, BM_CLICK, 0, 0);
     FlushMessages();
     COMPARE_CACHE(btnclick_nonthemed_sequence);    
 
     state = SendMessageW(hWnd2, BM_GETSTATE,0,0);
-    ok(state == 0, "Expected state 0, got %lu\n", state);
+    ok(state == 8, "Expected state 0, got %lu\n", state);
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply modules/rostests/apitests/comctl32/button.c]: Fixed
